wcwidth/wcwidth.py

```diff
@@ -190,19 +138,36 @@ def wcswidth(pwcs, n=None):
     or -1 if a non-printable character is encountered. When ``n`` is None
     (default), return the length of the entire string.
     """
-    # pylint: disable=C0103
-    #         Invalid argument name "n"
-
-    end = len(pwcs) if n is None else n
-    idx = slice(0, end)
     width = 0
-    for char in pwcs[idx]:
-        wcw = wcwidth(char)
-        if wcw < 0:
-            ucs = ord(char)
-            if _bisearch(ucs, NONZERO_COMBINING):
-                continue
-            return -1
-        else:
-            width += wcw
+    cache = {}
```

Comment:
A global cache statically populated with "a-zA-Z0-9" would still speed up english text processing and require neglegible memory. A well-populated static cache for east-asian languages would probably require a significant amount of memory ( O(10)KB ). A dynamic global cache would IMHO not be a very good idea, as in python `b'\xF0\xBF\xBF\xBF'.decode('utf-8')` decodes just fine to some very large, unused unicode codepoint and thus an attacker might attempt a memory exhaustion attack by spamming this cache with large, invalid code points.

A quick test in a python shell gives me about 50MB for a 1M entry width-dict, which at about 50byte/entry looks reasonable. That means such a cache could only be used with some culling strategy, and I'm not aware of a really fast way to do that to dicts in python. Actually, I think some form of size-limiting would be necessary even in case of the call-local cache to prevent this kind of attack. I'll do some tests on that later.

I think the call-local cache is quite neat in that it won't affect (already short) computation times for short strings by much, but it still speeds up computation time for long strings significantly while maintaining an OK memory footprint during the call.